vowpalwabbit/oaa.cc

```diff
@@ -85,11 +88,94 @@ void predict_or_learn(oaa& o, LEARNER::base_learner& base, example& ec)
     o.all->print_text(o.all->raw_prediction, outputStringStream.str(), ec.tag);
   }
 
-  ec.pred.multiclass = prediction;
+  if (is_probabilities)
+  { float sum_prob = 0;
+    for (uint32_t i=0; i<o.k; i++)
+    { // probability of class (i+1) = logistic_link_function(raw_prediction)
+      float prob = 1.f / (1.f + exp(- o.pred[i].scalar));
+      o.probs[i] = prob;
+      sum_prob += prob;
+    }
+    // make sure that the probabilities sum up (exactly) to one
+    for (uint32_t i=0; i<o.k; i++)
+      o.probs[i] /= sum_prob;
+    // copy probs pointer to the example.
+    // TODO: shouldn't we do a deep copy and free(ec.pred.probs) in finish_example()?
+    ec.pred.probs = o.probs;
+  } else {
+    ec.pred.multiclass = prediction;
+  }
+
   ec.l.multi = mc_label_data;
 }
 
-void finish(oaa&o) { free(o.pred); free(o.subsample_order); }
+void finish(oaa&o)
+{ free(o.pred);
+  free(o.subsample_order);
+  free(o.probs);
+}
+
+// TODO: partial code duplication with multiclass.cc:finish_example
+void finish_example_probabilities(vw& all, oaa& o, example& ec)
+{ // === Compute multiclass_log_loss
+  // TODO:
+  // What to do if the correct label is unknown, i.e. (uint32_t)-1?
```

Comment:
OK, but what should I do?
If I initialize `correct_class_prob = 1` (instead of 0, as it is now), I will get `multiclass_log_loss = 0` and thus effectively, I won't update `all.sd->multiclass_log_loss` nor `all.sd->holdout_multiclass_log_loss`. Should I change it this way?
I am still not sure, if I should increase `all.sd->holdout_multiclass_log_loss`.